Implement h-refinement projectors for volume data

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add transform_mutate for combining History objects

wthrowe committed:
Finish h-projection of Variables

wthrowe committed:
Implement h-projection of Tensor

wthrowe committed:
Implement h-projection of History

kidder commented on src/Time/History.hpp:
```diff
@@ -1121,4 +1123,86 @@ void transform(const gsl::not_null<History<DestVars>*> dest,
+                   << dest_record.time_step_id << " "
+                   << source_record.time_step_id);
+        ASSERT(dest_record.value.has_value() == source_record.value.has_value(),
+               "Only one entries to combine has a value at "
```
one entry or one of the entries

kidder commented on src/Time/History.hpp:
```diff
@@ -1121,4 +1123,86 @@ void transform(const gsl::not_null<History<DestVars>*> dest,
+  {
+    auto dest_substeps = dest->substeps();
+    const auto source_substeps = source.substeps();
+    ASSERT(dest->size() == source.size(),
```
shouldn't this be dest_substeps and source_substeps?

kidder replied:
test this and the above ASSERT

kidder commented on tests/Unit/ParallelAlgorithms/Amr/Projectors/Test_Variables.cpp:
```diff
@@ -93,35 +103,75 @@ void test_h_refine() {
+        children_data{};
+    children_data.insert(
+        {children_element_ids[0],
+         {children_elements[0], mesh, children_var_0[0], children_var_1[0]}});
```
test what happens if the children have different extents

kidder commented on tests/Unit/Evolution/Initialization/Test_Evolution.cpp:
```diff
@@ -325,6 +342,195 @@ void test_join() {
+          std::move(deriv0),          std::move(deriv1)};
+}
+
+void test_p_project_history() {
```
I was surprised there wasn't already a test for p-refinement; it appears I put the test into a separate file Test_TimeStepperHistory.  So either remove the separate file or move these tests there.

kidder commented on tests/Unit/Evolution/Initialization/Test_Evolution.cpp:
```diff
@@ -325,6 +342,195 @@ void test_join() {
+  }
+
+  {
+    auto box = db::create<db::AddSimpleTags<
```
test when the children have different extents

wthrowe committed:
Allow parent to be higher-resolution in projection

For the existing use case of boundary communication, the child
(mortar) always has the highest resolution, but for joining elements
in AMR the parent will be highest.

wthrowe commented:
Added several fixups and a new commit at the beginning: "Allow parent to be higher-resolution in projection".

kidder reviewed (commented):
Looks good. Squash

kidder commented:
ignoring clang-tidy timeout
